packages/network/src/connection/WebRtcEndpoint.ts

```diff
@@ -231,20 +227,108 @@ export class WebRtcEndpoint extends EventEmitter implements IWebRtcEndpoint {
+                connection.setDeferredConnectionAttempt(conn.stealDeferredConnectionAttempt())
+            }
+            
+            delete this.connections[peerId]
+            conn.close()
+
+            try {
+                connection.connect()
+            }
+            catch(e) {
+                this.logger.warn(e)
+            }
+            this.connections[peerId] = connection
+        }
+        else {
+            connection = this.connections[peerId]
+        }
+       
+        connection.setPeerInfo(PeerInfo.fromObject(originatorInfo))
+        connection.setConnectionId(connectionId)
+
+        connection.setRemoteDescription(description, 'offer' as DescriptionType.Offer)
+    }
+
+    async connect(
+        targetPeerId: string,
+        routerId: string,
+        trackerInstructed = true
+    ): Promise<string> {
+        const offering = this.peerInfo.peerId < targetPeerId
+        const peerType = offering ? 'offerer' : 'answerer'
+
+        // Prevent new connections from being opened when WebRtcEndpoint has been closed
+        if (this.stopped) {
+            return Promise.reject(new WebRtcError('WebRtcEndpoint has been stopped'))
+        }
+
+        if (this.connections[targetPeerId]) {
+            const connection = this.connections[targetPeerId]
+            const lastState = connection.getLastState()
+            this.logger.trace('%s has already connection for %s. state: %s', peerType, NameDirectory.getName(targetPeerId), lastState)
+            if (lastState === 'connected') {
+                return Promise.resolve(targetPeerId)
+            }
+
+            if (connection.getDeferredConnectionAttempt()) {
+                return (connection.getDeferredConnectionAttempt()?.connectionAttemptPromise as Promise<string>)
+            }
+            else {
+                throw new Error(`disconnected ${connection.getPeerId()}`)
+            }
+        }
+        
+        const connection = this.createConnection(targetPeerId, routerId, offering)
+        connection.createDeferredConnectionAttempt()
+
+        if (offering) {
+            connection.setConnectionId(uuidv4())
+        }
 
         this.connections[targetPeerId] = connection
         connection.connect()
-        if (!isOffering) {
+        
+        if (!trackerInstructed && !offering) {
             this.rtcSignaller.onConnectionNeeded(routerId, connection.getPeerId())
         }
 
-        await Promise.race([
-            once(connection, 'open'),
-            once(connection, 'close').then(() => {
-                throw new Error(`disconnected ${connection.getPeerId()}`)
-            }),
-        ])
-        return connection.getPeerId()
+        if (connection.getDeferredConnectionAttempt()) {
+            return (connection.getDeferredConnectionAttempt()?.connectionAttemptPromise as Promise<string>)
+        }
+        else {
+            throw new Error(`disconnected ${connection.getPeerId()}`)
```

Comment:
This should be more or less impossible though, since we invoked `connection.createDeferredConnectionAttempt()` earlier? If this were nullish, it would be indicative of a bug in our code? So edited the error message to reflect that:
```suggestion
            throw new Error(`unexpected deferedConnectionAttempt == null ${connection.getPeerId()}`)
```

Reply:
ok